certified-connectors/DocuSignDemo/script.csx

```diff
@@ -2118,13 +2118,90 @@ public class Script : ScriptBase
         signer = new JObject();
       }
+    }
+
+    newBody["templateRoles"] = templateRoles;
+
+    if (!string.IsNullOrEmpty(query.Get("status")))
+    {
+      newBody["status"] = query.Get("status");
+    }
 
-      if (key.Contains(" Signing Group"))
+    return newBody;
+  }
+
+  private JObject CreateEnvelopeFromTemplateV3BodyTransformation(JObject body)
+  {
+    var templateRoles = new JArray();
+    var query = HttpUtility.ParseQueryString(this.Context.Request.RequestUri.Query);
+
+    var uriBuilder = new UriBuilder(this.Context.Request.RequestUri);
+    uriBuilder.Path = uriBuilder.Path.Replace("/envelopes/createWithRecipientFields", "/envelopes");
+    this.Context.Request.RequestUri = uriBuilder.Uri;
+
+    var newBody = new JObject()
+    {
+      ["templateId"] = query.Get("templateId"),
+      ["emailSubject"] = query.Get("emailSubject"),
+      ["emailBlurb"] = query.Get("emailBody")
+    };
+
+    Dictionary<string, JObject> recipientMapping = new Dictionary<string, JObject>();
+    foreach (var property in body)
+    {
+      var value = (string)property.Value;
+      var key = (string)property.Key;
+      string[] keyArray = key.Split(new string[]{":::"}, StringSplitOptions.None);
+      var roleName = keyArray[0];
+      // custom fields parsing to match request body object from Docusign API
+      if (string.Equals(keyArray[0], "List Custom Fields", StringComparison.OrdinalIgnoreCase))
       {
-        signer["signingGroupId"] = value;
-        templateRoles.Add(signer);
-        signer = new JObject();
+        if (!newBody.ContainsKey("customFields"))
+        {
+          newBody["customFields"] = new JObject();
+        }
+        JObject recipientCustomFieldObj = (JObject) newBody["customFields"];
+        if (!recipientCustomFieldObj.ContainsKey("listCustomFields"))
+        {
+          recipientCustomFieldObj["listCustomFields"] = new JArray();
+        }
+        JArray listCustomFieldsArray = (JArray) recipientCustomFieldObj["listCustomFields"];
+        listCustomFieldsArray.Add(new JObject
+        {
+          ["name"] = keyArray[1],
+          ["value"] = value,
+          ["show"] = "true"
+        });
+        continue;
       }
+      if (string.Equals(keyArray[0], "Text Custom Fields", StringComparison.OrdinalIgnoreCase))
+      {
+        if (!newBody.ContainsKey("customFields"))
+        {
+          newBody["customFields"] = new JObject();
+        }
+        JObject recipientCustomFieldObj = (JObject) newBody["customFields"];
+        if (!recipientCustomFieldObj.ContainsKey("textCustomFields"))
+        {
+          recipientCustomFieldObj["textCustomFields"] = new JArray();
+        }
+        JArray textCustomFieldsArray = (JArray) recipientCustomFieldObj["textCustomFields"];
+        textCustomFieldsArray.Add(new JObject
+        {
+          ["name"] = keyArray[1],
+          ["value"] = value,
+          ["show"] = "true"
+        });
+        continue;
+      }
```

Comment:
There is code repeating from line 2159 to 2196. You may want to pull the common code in a method.

Reply:
TEst

Reply:
One of the fields is "textCustomFields" and the other one is for "listCustomFields". So even if we were to refactor and write a private function, the lines of code would not be significantly reduced.

Also, it would be risky to update this part of the code as it dynamically generates fields that, if changed, would break backwards compatibility.